### PR TITLE
Openstack cloud tests

### PR DESCRIPTION
**Pull request by mkalcok**
Please review commit by commit. This branch is based on #175 so the first 7 commits by @fnordahl can be skipped in review.

This change builds on separation of CMS implementations in #175 and it adds Openstack CMS. More info is in commit messages but at current stage it can simulate provisioning of:

- Project router
- Project's Internal network
- External network with gateway routing

### CODE REVIEW

File: ovn-tester/ovn_utils.py
Comment (fnordahl) [inline ovn-tester/ovn_utils.py:435]: `enabled: bool = False`

This changes the behavior of the method and is the cause of the CI functional test failure.

Should this either be `enabled: Optional[bool] = None` or `enabled: bool = True`?

File: ovn-tester/cms/openstack/openstack.py
Comment (fnordahl) [inline ovn-tester/cms/openstack/openstack.py:91]: Note that I think many OpenStack clouds default to re-using the same CIDR for the internal network of every project.

Using different ones makes it simpler to have overview in a test though and allows for testing both NAT'ed and routed use cases.

I guess we need to consider whether having many networks using the same CIDR impacts scale and whether we need to test that somewhere down the line.

File: ovn-tester/cms/openstack/openstack.py
Comment (fnordahl) [inline ovn-tester/cms/openstack/openstack.py:275]: The settings for the port in this method looks like they come from the special Neutron OVN Metadata port.

For a regular VM port the type would be `""` and the `requested-chassis` would point at an actual node.

**Review by fnordahl — COMMENTED** [review]
Thanks a lot for your work on this, it looks like a great start! I have left a couple of comments in-line and I'll continue to look later today

File: ovn-tester/cms/openstack/openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/openstack.py:275]: I think I just misunderstood the port that was created and thus I named it wrong. When user runs `openstack network create`, Neutron automatically creates one `LSPort` for the `LSwitch`. I wasn't entirely clear on the purposed of this port so I gave it this general name/description. Do you have any suggestion on naming/purpose of this port?

File: ovn-tester/ovn_utils.py
Comment (mkalcok) [reply ovn-tester/ovn_utils.py:435]: Thanks for the pointer, you are right. Without passing the `enabled` field, the router has "empty" value in the column which I mistook for the same thing as `enabled=false`

Comment (mkalcok) [reply ovn-tester/ovn_utils.py:435]: So I found a [source on this behavior](https://man7.org/linux/man-pages/man5/ovn-nb.5.html#Logical_Router_TABLE). empty/unset value in `enabled` column is equal to `true`.
Even though the Neutron sets the value to `true` explicitly, I'd say we can drop this change and let it be the default "unset".

File: ovn-tester/cms/openstack/openstack.py
Comment (rjarry) [inline ovn-tester/cms/openstack/openstack.py:297]: This is a general remark. It looks like you are not using transactions, whereas the ML2 driver tries to do so as much as possible. I'm not sure if it has any performance implications. If not, maybe that's not important.

File: ovn-tester/cms/openstack/openstack.py
Comment (rjarry) [inline ovn-tester/cms/openstack/openstack.py:305]: At first, I though this was a typo, but it is handled as a special case. You pass it as a string and it is parsed as a `dict` internally. Weird...

**Review by rjarry — COMMENTED** [review]
Hi,

thanks for working on this. I had gone another route (trying to mock some parts of neutron to reuse the actual ML2 driver code) which has proven unfruitful.

I don't see anything about the metadata agent here nor any southbound access to simulate what the ML2 driver does. Is there any plan to do so?

File: ovn-tester/cms/openstack/openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/openstack.py:305]: yeah, it's a weird inconsistency :D

File: ovn-tester/cms/openstack/openstack.py
Comment (fnordahl) [reply ovn-tester/cms/openstack/openstack.py:91]: Agreed, I think your approach is the best to start with as it is easier to work with and covers multiple use cases. If we come to the conclusion that creating multiple networks with the same CIDR is an important use case we can change it later.

File: ovn-tester/cms/openstack/openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/openstack.py:297]: Thanks for the tip about the transaction, I did not consider them. I assume that we would not be measuring/profiling the "cluster_bringup" but better performance never hurts. I'll look into where I could group calls into transactions :+1:

**Comment by mkalcok** [comment]
Thanks for the reviews @fnordahl @rjarry.

> I don't see anything about the metadata agent here nor any southbound access to simulate what the ML2 driver does. Is there any plan to do so?

That's a good point. I focused on the NB because I assumed that that would be the primary database the Neutron would communicate with. I'll take a look at the SB as well. Do you have any pointers to relevant code/docs about this?

**Comment by rjarry** [comment]
There are a two parts to consider:

1) the neutron server access to the southbound db: https://opendev.org/openstack/neutron/src/tag/21.1.2/neutron/scheduler/l3_ovn_scheduler.py

2) the metadata agent that runs on all compute nodes: https://opendev.org/openstack/neutron/src/tag/21.1.2/neutron/agent/ovn/metadata/agent.py

I am far from being an expert in this code but I know it has significant impact on scale because it multiplies the number of southbound db clients by two.

For performance reasons, openstack deployments usually recommend enabling ovn-controller `monitor_all=true` to reduce cpu pressure on the southbound ovsdb-server. The same logic is also applied for neutron metadata agent. To properly do scale testing, it would be good for ovn-heater to simulate these connections and subscriptions.

**Comment by dceara** [comment]
> There are a two parts to consider:
> 
>     1. the neutron server access to the southbound db: https://opendev.org/openstack/neutron/src/tag/21.1.2/neutron/scheduler/l3_ovn_scheduler.py
> 
>     2. the metadata agent that runs on all compute nodes: https://opendev.org/openstack/neutron/src/tag/21.1.2/neutron/agent/ovn/metadata/agent.py
> 
> 
> I am far from being an expert in this code but I know it has significant impact on scale because it multiplies the number of southbound db clients by two (1 ovn-controller + 1 neutron metadata agent per compute node).
> 
> For performance reasons, openstack deployments usually recommend enabling ovn-controller `monitor_all=true` to reduce cpu pressure on the southbound ovsdb-server. The same logic is also applied for neutron metadata agent. To properly do scale testing, it would be good for ovn-heater to simulate these connections and subscriptions.

It seems to me we're only interested in how these extra read-only clients increase load on the Southbound.  If that's the case, we can simulate DB clients by running ovsdb-server relay instances.  Today we already start real SB relays if configured ([e.g., `cluster.n_relays > 0`](https://github.com/ovn-org/ovn-heater/blob/f19d060f8ea54650b8868272977aa345229d04a9/test-scenarios/ovn-low-scale.yml#L11)).  We can programmatically start more of these (without connecting ovn-controllers to them) and achieve the same type of load on the SB as with neutron agents.   IIRC @felixhuettner had also suggested this during one of the first ovn-heater community meetings.

**Comment by felixhuettner** [comment]
> It seems to me we're only interested in how these extra read-only clients increase load on the Southbound. If that's the case, we can simulate DB clients by running ovsdb-server relay instances. Today we already start real SB relays if configured ([e.g., `cluster.n_relays > 0`](https://github.com/ovn-org/ovn-heater/blob/f19d060f8ea54650b8868272977aa345229d04a9/test-scenarios/ovn-low-scale.yml#L11)). We can programmatically start more of these (without connecting ovn-controllers to them) and achieve the same type of load on the SB as with neutron agents. IIRC @felixhuettner had also suggested this during one of the first ovn-heater community meetings.

I think that should give us a fairly accurate measurement of the neutron-ovn-metadata-agent. afaik it currently does not set monitor conditions, so we get a full db dump anyway (currently clarifying that).
The only thing we are loosing is the performance difference between a relay and the neutron agent when receiving updates from the ovsdb; but i don't think that should matter too much

File: ovn-tester/cms/openstack/tests/base_openstack.py
Comment (fnordahl) [inline ovn-tester/cms/openstack/tests/base_openstack.py:28]: I like this, it uses the same approach as the `base_clustrer_bringup` test that every ovn-kubernetes test case runs to initialize things before running other tests.

The `new_project` method should probably be called N times per test scenario where N is input from the test-scenario configuration without direct relation to the number of workers configured. Perhaps a `n_projects` key under `base_openstack`?

About the number of `gw_nodes`. Neutron currently has a hard coded maximum of 5 GW nodes per router, so we should probably mimic that behavior. The nodes are either selected by chance or by load (load as in number of routers hosted at any given priority)

**Comment by felixhuettner** [comment]
> @dceara: I think the neutron metadata agent is not read-only, it also writes to the southbound db:
> 
> https://opendev.org/openstack/neutron/src/tag/21.1.2/neutron/agent/ovn/metadata/agent.py#L193-L196
> 
> https://opendev.org/openstack/neutron/src/tag/21.1.2/neutron/agent/ovn/metadata/agent.py#L292-L293

Yep,
there is also some regular heartbeat, but most people configure it 

> > It seems to me we're only interested in how these extra read-only clients increase load on the Southbound. If that's the case, we can simulate DB clients by running ovsdb-server relay instances. Today we already start real SB relays if configured ([e.g., `cluster.n_relays > 0`](https://github.com/ovn-org/ovn-heater/blob/f19d060f8ea54650b8868272977aa345229d04a9/test-scenarios/ovn-low-scale.yml#L11)). We can programmatically start more of these (without connecting ovn-controllers to them) and achieve the same type of load on the SB as with neutron agents. IIRC @felixhuettner had also suggested this during one of the first ovn-heater community meetings.
> 
> I think that should give us a fairly accurate measurement of the neutron-ovn-metadata-agent. it currently does not set monitor conditions, so we get a full db dump anyway. The only thing we are loosing is the performance difference between a relay and the neutron agent when receiving updates from the ovsdb; but i don't think that should matter too much

Allthough it seems like there is now some kind of monitor conditions going on here, but only for chassis https://opendev.org/openstack/neutron/src/tag/21.1.2/neutron/agent/ovn/metadata/ovsdb.py#L50-L53

File: ovn-tester/cms/openstack/tests/base_openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/tests/base_openstack.py:28]: That's a good point about the GW nodes, thank you. I didn't know about the gw limit per router. It should be fairly easy to distribute chassis in balanced way

File: ovn-tester/cms/openstack/openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/openstack.py:297]: I've looked into this bit more today and I don't see a nice way to integrate transactions. Currently I'm using mainly `ovn_utils.OvnNbctl` moethods which are not "batch-able". I could get around it by using `OvnNbctl.idl` directly but then we'd lose code reusing.
Also, correct me if I'm wrong here, but ovn-heater does not measure/profile "cluster_bringup", does it? If not, I'd lean towards not implementing transactions now.

File: ovn-tester/ovn_utils.py
Comment (mkalcok) [reply ovn-tester/ovn_utils.py]: I've dropped the `enabled` parameter entirely, so I'll mark this conversation as resolved.

File: ovn-tester/cms/openstack/openstack.py
Comment (rjarry) [reply ovn-tester/cms/openstack/openstack.py:336]: I don't know if cluster bringup is measured by ovn-heater. But that's ok if we don't use transactions. At worst we'll get lower results. Since CMS's most likely use them, performance can only improve :)

File: ovn-tester/cms/openstack/tests/base_openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/tests/base_openstack.py]: I had to eat my words about the easy redistribution mechanism :D balance distribution of individual GW priorities between bunch of routers turned out rather complex so I defaulted to using `random.sample()` :D

File: ovn-tester/cms/openstack/openstack.py
Comment (felixhuettner) [inline ovn-tester/cms/openstack/openstack.py:151]: From my experience openstack environments have generally a low count of widely used external networks (sometimes just one). So i would here not generate a new external network, but rather randomly choose from a list of existing ones.

File: ovn-tester/cms/openstack/openstack.py
Comment (felixhuettner) [inline ovn-tester/cms/openstack/openstack.py:153]: as far as i know there will only ever be one provider network port per provider network. Since `_add_provider_network_port` sets this statically to `physnet1` it should only ever be called once

File: ovn-tester/cms/openstack/openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/openstack.py:151]: These are two very good points thank you. I'll decouple ext_net creation from project creation.

**Comment by mkalcok** [comment]
@felixhuettner I decoupled external network creation from project creation and now the `BaseOpenstack` test creates only one ext_net and connects all projects to it.

I'm creating these suggested changes as separate commits for easier review. Please let me know if that's OK or if you'd like me to squash them in the end to [53339f9](https://github.com/ovn-org/ovn-heater/pull/179/commits/53339f9f79e8ba202a3e3d2fecbb54608243e08c)

**Comment by felixhuettner** [comment]
> @felixhuettner I decoupled external network creation from project creation and now the `BaseOpenstack` test creates only one ext_net and connects all projects to it.
> 
> I'm creating these suggested changes as separate commits for easier review. Please let me know if that's OK or if you'd like me to squash them in the end to [53339f9](https://github.com/ovn-org/ovn-heater/pull/179/commits/53339f9f79e8ba202a3e3d2fecbb54608243e08c)

Thanks. Looks good

**Comment by mkalcok** [comment]
I'd say that this change is now almost ready for full review. Two things that are missing and I'd like your opinion on are:

- Floating IPs and external connectivity: I'd be happy to do this in separate PR to not blow up scope of this one even more.
- ovn-metadata-agents: Is the consensus that we can simulate metadata agents with ovn-relay nodes to create "read" load on SB DB? Since as far as I can tell the only write operations of metadata agents are keepalive messages and updates to chassis external ID's on registration?

**Comment by fnordahl** [comment]
> I'd say that this change is now almost ready for full review. Two things that are missing and I'd like your opinion on are:
> 
>     * Floating IPs and external connectivity: I'd be happy to do this in separate PR to not blow up scope of this one even more.
> 
>     * ovn-metadata-agents: Is the consensus that we can simulate metadata agents with ovn-relay nodes to create "read" load on SB DB? Since as far as I can tell the only write operations of metadata agents are keepalive messages and updates to chassis external ID's on registration?

From my perspective I think it would be valuable to finalize this PR as-is now and mark it as ready for review once you feel comfortable with it.

The items discussed above can be worked on independently, and the work you have done so far would be a dependency.

**Comment by felixhuettner** [comment]
> @felixhuettner I decoupled external network creation from project creation and now the `BaseOpenstack` test creates only one ext_net and connects all projects to it.
> 
> I'm creating these suggested changes as separate commits for easier review. Please let me know if that's OK or if you'd like me to squash them in the end to [53339f9](https://github.com/ovn-org/ovn-heater/pull/179/commits/53339f9f79e8ba202a3e3d2fecbb54608243e08c)

I'm not sure what has happend, but we seem to create multiple external networks again (even though they now seem to have no external connectivity)?

**Comment by mkalcok** [comment]
:facepalm: thank you for noticing. I lost a commit that decoupled project/ext net creation when I was rebasing/force pushing local copy on a different machine

**Review by felixhuettner — APPROVED** [review]

**Review by rjarry — APPROVED** [review]

File: ovn-tester/cms/openstack/tests/base_openstack.py
Comment (fnordahl) [inline ovn-tester/cms/openstack/tests/base_openstack.py]: I wonder if we should find a more specific name for this configuration option. While it fits well with the current implementation, it does not really fit with how it could be used in OVN.

The `Gateway_Chassis` (or possibly `HA_Chassis_Group` at some point in the future) is a property of the [`Logical_Router_Port`](https://manpages.ubuntu.com/manpages/jammy/man5/ovn-nb.5.html#logical_router_port%20table).

A project can have multiple routers, and routers can have multiple logical router ports with different configuration for these attributes. There is [ongoing work in OpenStack to support multiple LRPs per router](https://specs.openstack.org/openstack/neutron-specs/specs/2023.2/active-active-l3-gateway-with-multihoming.html).

Would perhaps `n_chassis_per_gw_lrp` work?

File: test-scenarios/openstack.yml
Comment (fnordahl) [inline test-scenarios/openstack.yml]: nit: newline at end of file please

**Review by fnordahl — COMMENTED** [review]
Thank you for all your work on this, the proposal gives us a solid foundation (in record time) on which multiple people can iterate independently to implement the myriad of choices in OpenStack networking.

I have one comment about choice of configuration option name that I think we should address and a nit in-line.

We should probably also add this test case as a Cirrus CI job but I think that can be done as a follow-up.

Otherwise, this LGTM!

File: ovn-tester/cms/openstack/tests/base_openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/tests/base_openstack.py]: hmm. If we call it `n_chassis_per_gw_rlp`, do we also need to tie it back to the projects with something like `n_gw_ports_per_project` ?

Comment (fnordahl) [reply ovn-tester/cms/openstack/tests/base_openstack.py]: > hmm. If we call it `n_chassis_per_gw_rlp`, do we also need to tie it back to the projects with something like `n_gw_ports_per_project` ?

I'm very picky on configuration option naming, as opposed to implementations which change frequently, options are a kind of API and have a infinite lifetime.  Any future changes to their name or meaning may have expensive workflow change costs down the line.

So my intention with this comment was scoped at this specific option.

But you are right in that it would probably also make sense to have other options to control the adjacent entities in the project-router-lrp test matrix.

Would something like `n_gw_ports_per_router` and `n_routers_per_project` catch more use cases?

Not sure if we need them right now though.

Comment (mkalcok) [reply ovn-tester/cms/openstack/tests/base_openstack.py]: You are right. We need to strike balance between future-proofing and not inventing unnecessary options here. We can stick with `n_chassis_per_gw_lrp` with implication that currently each project has only one router and one gw port.

**Comment by mkalcok** [comment]
I pushed two small additions:

- Rename config option parameter according to @fnordahl suggestion
- Added task to `.cirrus.yml` to run low-scale Openstack test as part of the gate

**Review by felixhuettner — APPROVED** [review]

**Comment by fnordahl** [comment]
I'm working on a rebase after #180 merged.

**Comment by fnordahl** [comment]
Rebased and ready for review now @dceara @felixhuettner @rjarry

File: ovn-tester/cms/openstack/openstack.py
Comment (booxter) [inline ovn-tester/cms/openstack/openstack.py:446]: This was removed lately https://review.opendev.org/c/openstack/neutron/+/888127

File: ovn-tester/cms/openstack/openstack.py
Comment (booxter) [inline ovn-tester/cms/openstack/openstack.py:488]: I will remove the empty chassis key in ml2/ovn, it's not needed: https://review.opendev.org/c/openstack/neutron/+/897489

File: ovn-tester/cms/openstack/openstack.py
Comment (booxter) [inline ovn-tester/cms/openstack/openstack.py:487]: (drive-by comment) these endpoints should probably be renamed to avoid double _set_.

File: ovn-tester/cms/openstack/openstack.py
Comment (booxter) [inline ovn-tester/cms/openstack/openstack.py:579]: (As a matter to discuss) A lot of effort here (passing parameters with "real" values, creating the dict) is spent to populate ext_ids with keys that are identical to a "real" openstack. But these keys are opaque for OVN, so we could shave off some complexity by either a) omitting populating the keys at all or b) if we want to make sure the db is populated with something, maybe just generate random strings / hardcode.

File: ovn-tester/cms/openstack/openstack.py
Comment (booxter) [inline ovn-tester/cms/openstack/openstack.py:304]: we may want to also set mtu_request, as os-vif ovs port implementation does. This may affect behavior of OVN in some cases. (but this can be left to later.)

https://github.com/openstack/os-vif/blob/0a0dec37e42e220ffa750650027714bd7289faa1/vif_plug_ovs/ovsdb/ovsdb_lib.py#L68

File: ovn-tester/cms/openstack/openstack.py
Comment (fnordahl) [reply ovn-tester/cms/openstack/openstack.py:579]: We did ask ourselves whether it was worth populating them, and we concluded with three reasons to keep them:
1. To our surprise we found they are not entirely opaque, and there is actually processing of them in OVN [0][1].
2. Neutron's widespread use of external_ids increases the size of the database, we ought to measure what impact this has.
3. If you are developing/debugging the ovn-heater OpenStack plugin with a Neutron background, having these familiar keys with data that actually link to something can be useful.

0: https://github.com/ovn-org/ovn/blob/430c0fd32dd709b8d452ce9a339dd30228744b62/northd/northd.c#L1114-L1120
1: https://github.com/ovn-org/ovn/blob/430c0fd32dd709b8d452ce9a339dd30228744b62/northd/northd.c#L3683-L3690

File: ovn-tester/ovn_workload.py
Comment (dceara) [inline ovn-tester/ovn_workload.py:391]: Is it worth accounting(`@ovn_stats.timeit`) how long the mesh_ping took in total?

File: ovn-tester/cms/openstack/openstack.py
Comment (fnordahl) [reply ovn-tester/cms/openstack/openstack.py:487]: I agree that looks weird, I would like to propose we address that pre-existing issue in `ovn-tester/ovn_utils.py` separately though.

**Comment by fnordahl** [comment]
@booxter @dceara for when you have a spare moment, I believe all review feedback has been addressed.

**Review by felixhuettner — APPROVED** [review]

**Review by booxter — APPROVED** [review]
Code wise I don't have any particular issues. Also, I believe we can iterate on it later and it's better to merge this PR that is already quite huge.

(I assume it doesn't break existing k8s scenarios.)

**Comment by dceara** [comment]
@mkalcok @fnordahl There are a few conflicts due to merging the initial ovn-ic work; I'll try to help with the rebase.

**Comment by dceara** [comment]
I tried a rebase here (completely untested yet): https://github.com/dceara/ovn-heater-1/commits/openstack-cloud-setup-rebase-tmp

**Comment by fnordahl** [comment]
> I tried a rebase here (~completely untested yet~ barely tested): https://github.com/dceara/ovn-heater-1/commits/openstack-cloud-setup-rebase-tmp

Thanks, I'll have a look

**Comment by fnordahl** [comment]
@dceara thanks a lot for the help with the rebase. I put your proposals for adaption into the correct commits and attributed accordingly.

I redid the parts of the commits that mainly move code around to ensure no changes were lost.

I also arrived at a slightly different solution to calling the `prepare_test` method, hope you agree.

Looking forward to your review!

**Comment by fnordahl** [comment]
I just noticed that the rebase is one commit shorter than the previous iteration, and the reason is that I have unintentionally squashed "CMS Openstack: Add ability to simulate VMs" and "OS Test: Spawn VMs and run ping in base scenario".

However, the result does look fine to me as both commits kind of addressed the same thing in iterations, so perhaps we should just keep it this way? Thoughts @mkalcok ?

**Comment by mkalcok** [comment]
yeah, I'm fine with them being squashed. I separated them for easier review as one added logic to the `CMS Openstack` implementation and the other used that implementation in a test scenario, but they are closely related.

File: ovn-tester/ovn_utils.py
Comment (dceara) [inline ovn-tester/ovn_utils.py:5]: Thanks a lot for adding the type hints!

File: ovn-tester/cms/openstack/openstack.py
Comment (dceara) [inline ovn-tester/cms/openstack/openstack.py:121]: We had some discussions internally about potentially removing the ovn-fake-multinode dependency and start OVN components directly from ovn-heater.  Whenever (if) we do that it might be a good moment for adding support for network nodes too.

File: ovn-tester/cms/openstack/openstack.py
Comment (dceara) [inline ovn-tester/cms/openstack/openstack.py:316]: Nit: I understand the intent but on the other hand, this is a testing tool and reproducible runs are very important .  Not necessarily in this PR but do we need to log the chosen values and provide a way to explicitly select the same chassis if we want to re-run the exact same workload while debugging a DUT issue?

File: ovn-tester/ovn_utils.py
Comment (dceara) [inline ovn-tester/ovn_utils.py:472]: We're missing:
```
options = {} if options is None else options
```

which causes:
https://cirrus-ci.com/task/5461921055375360?logs=test_openstack#L447
```
2023-10-25 11:40:24,426 | ovsdbapp.backend.ovs_idl.vlog |ERROR| attempting to write bad value to column options (ovsdb error: expected string, got <class 'NoneType'>)
```

@mkalcok I can squash it in if you agree with this change.

**Review by dceara — APPROVED** [review]
This is awesome work!  Thanks a lot @mkalcok @fnordahl @booxter @rjarry @felixhuettner for all the code and reviews!

I agree with @booxter let's get this in and we can address leftovers/known issues/bugs with follow up PRs.

There is one thing I'd like to fix first before merging this but I'll wait for a reply from @mkalcok or @fnordahl on https://github.com/ovn-org/ovn-heater/pull/179#pullrequestreview-1700628437

Thanks,
Dumitru

File: ovn-tester/cms/openstack/openstack.py
Comment (mkalcok) [reply ovn-tester/cms/openstack/openstack.py:316]: This is a very good point. I originally intended to do even distribution of GWs but the implementation turned out more complicated than I anticipated so I went with random. I'll open a ticket about this so that the issue is not lost.

File: ovn-tester/ovn_utils.py
Comment (mkalcok) [reply ovn-tester/ovn_utils.py:472]: Yes, thank you for noticing this. I seem to remember fixing this one already, not sure how it resurfaced. I'll take care of it.

Comment (fnordahl) [reply ovn-tester/ovn_utils.py:472]: > Yes, thank you for noticing this. I seem to remember fixing this one already, not sure how it resurfaced. I'll take care of it.

No wonder you thought that, because apparently I regressed this in "cms/openstack: Set MTU for VM and GW ports."! Sorry about that and thank you for fixing.

Comment (dceara) [reply ovn-tester/ovn_utils.py:472]: Thanks, looks good now!

File: ovn-tester/cms/openstack/openstack.py
Comment (dceara) [reply ovn-tester/cms/openstack/openstack.py:316]: Thanks @mkalcok

**Pull request merged**
